### PR TITLE
Ability to resume queue asynchronously, chaining, more events

### PR DESCRIPTION
**Pull request by IvanProdaiko94**
Add:
- Ability to resume queue asynchronously;
- Chaining;
With this pull request one can resume execution asynchronously;

### CODE REVIEW

File: lib/queue.js [hunk 1]
```diff
@@ -21,45 +21,57 @@ metasync.ConcurrentQueue = function(
   };
 };
 
+
+const process = function () {
```
Comment (aqrln) [inline lib/queue.js]: There's no space between `function` keyword and parentheses according to our codestyle. I'd prefer it to be a named function, though.

File: lib/queue.js [hunk 2]
```diff
@@ -21,45 +21,57 @@ metasync.ConcurrentQueue = function(
...
+  if (!this.isOnPause) {
+    if (this.count < this.concurrency) {
+      let item = this.items.shift();
```
Comment (aqrln) [inline lib/queue.js]: `const`

**Comment by aqrln** [comment]
@IvanProdaiko94 why did you close the PR? :)

**Comment by IvanProdaiko94** [comment]
Just find out the I have done smth wrong. I will open it just after improve it;

File: lib/queue.js [hunk 3]
```diff
@@ -115,4 +125,4 @@ metasync.ConcurrentQueue.prototype.stop = function() {
     process: null
   };
   return this;
-};
+};
```
Comment (aqrln) [inline lib/queue.js]: No newline at the end of file

File: lib/queue.js [hunk 4]
```diff
@@ -107,4 +124,5 @@ metasync.ConcurrentQueue.prototype.stop = function() {
     empty: null,
     process: null
   };
-};
+  return this;
+};
```
Comment (aqrln) [inline lib/queue.js]: No newline at the end of file

**Comment by IvanProdaiko94** [comment]
The main idea is that when I tried to execute this code 
```
let q = new metasync.ConcurrentQueue(3, 2000);
q.on('process', (item, cb) => {
  setTimeout(() => cb(), 500)
});
for (let i = 0; i <= 50; i++) {
  q.add({id : i});
}
q.pause();
setTimeout(() => q.resume(), 2500);
```
it didn't resume asynchronously.
It happened because `resume` method just change the `isOnPause` property. To my mind it should force queue to start again.

**Comment by aqrln** [comment]
I'm not sure I got your point right but if you mean that it didn't continue execution of workers that had asynchronous operations, it sounds reasonable.

**Comment by aqrln** [comment]
Some nits:

* Can you please format your commit messages according to common Git guidelines so that `git log` is readable? I don't think we have strong commit conventions like mandatory 50/72 wrapping in this repo but at least they should look like Git commit messages that don't make Git, GitHub and common CLI tools output a bunch of `Add:` lines without any extra information :)

  Some tips:

  <http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html>
  <https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message>

* Rebase your commits against `master` so that there's no junk merge commit.

* I'd prefer the last four commits to be squashed so that there are only two commits in this PR: one that adds chaining and one that fixes a bug.

**Comment by aqrln** [comment]
Aside from that, it would be nice to include tests together with the patch.

File: lib/queue.js [hunk 5]
```diff
@@ -31,35 +31,40 @@ metasync.ConcurrentQueue.prototype.add = function(
...
-    timer = setTimeout(() => {
-      const err = new Error('ConcurrentQueue timed out');
-      queue.emit('timeout', err);
-    }, queue.timeout);
-  }
-  const stub = (item, callback) => callback();
-  const fn = queue.events.process || stub;
-  fn(item, () => {
-    queue.count--;
+  if (!queue.isOnPause) {
```
Comment (tshemsedinov) [inline lib/queue.js]: It is better to immediate return here:
```js
if (queue.isOnPause) return this;
```

Comment (tshemsedinov) [inline lib/queue.js]: this will be goog for code style and this will minimize changes so PR purpose will be clear

Comment (IvanProdaiko94) [inline lib/queue.js]: Ok. I will do this.

File: lib/queue.js [hunk 6]
```diff
@@ -31,14 +31,15 @@ metasync.ConcurrentQueue.prototype.add = function(
...
+  return this;
 };
 
 metasync.ConcurrentQueue.prototype.next = function(
   item // process next item from queue
 ) {
   const queue = this;
   let timer;
-  if (queue.isOnPause) return;
+  if (!queue.isOnPause) return this;
```
Comment (tshemsedinov) [inline lib/queue.js]: Remove !
```js
if (queue.isOnPause) return this;
```

File: lib/queue.js [hunk 7]
```diff
@@ -85,14 +90,25 @@ metasync.ConcurrentQueue.prototype.emit = function(
...
 metasync.ConcurrentQueue.prototype.pause = function() {
   this.isOnPause = true;
+  return this;
 };
 
 metasync.ConcurrentQueue.prototype.resume = function() {
   this.isOnPause = false;
+  while (this.count < this.concurrency) {
+    const item = this.items.shift();
+    if (item) {
```
Comment (tshemsedinov) [inline lib/queue.js]: Here we may shorten:
```js
if (!item) return this;
this.next(item);
```

Comment (IvanProdaiko94) [inline lib/queue.js]: Ok

**Comment by aqrln** [comment]
@tshemsedinov does this LGTY after @IvanProdaiko94 has pushed a couple of new commits?

**Comment by aqrln** [comment]
Rebased onto current master, fixed the conflicts, commit messages and stylistic issues in https://github.com/metarhia/MetaSync/tree/pr-46-rebased.

CI: https://travis-ci.org/metarhia/MetaSync/builds/215359722

**Comment by aqrln** [comment]
Landed on master in 25444073c5f854f43c9a2f1124391edfb9b7c142, 73feef8d86b16471d20a7558ab2243f4edec9d4b and 742288580a45d4455465ce2fe263029d74a6b2f1. Thanks for the contribution and sorry for the PR being lost and forgotten.

**Pull request closed without merging**